app/src/main/java/com/android/mygarden/ui/garden/GardenViewModel.kt

```
@@ -35,35 +48,70 @@ class GardenViewModel(
+   * Sets the error message of the UI state.
    *
    * @param msg the error message
    */
   fun setErrorMsg(msg: String) {
     _uiState.value = _uiState.value.copy(errorMsg = msg)
   }
 
-  /** Fetches all the plants from the [plantsRepo] and set an error message if the fetch failed */
-  fun getAllPlants() {
+  /**
+   * Wrapper for the repository homologous function that update the given plant last time watered to
+   * the current time of the device.
+   *
+   * @param ownedPlant the plant that is being watered
+   */
+  fun waterPlant(ownedPlant: OwnedPlant) {
+    viewModelScope.launch {
+      plantsRepo.waterPlant(ownedPlant.id, Timestamp(System.currentTimeMillis()))
+      // Refresh UI state
+      // TODO: update when the health status update is implemented
+      getAllPlants()
+    }
+  }
+
+  /** Fetches all the plants from the [plantsRepo] or set an error message if the fetch failed. */
+  private fun getAllPlants() {
     viewModelScope.launch {
       try {
         val plants = plantsRepo.getAllOwnedPlants()
-        _uiState.value = GardenUIState(plants = plants)
+        _uiState.value = _uiState.value.copy(plants = plants)
       } catch (e: Exception) {
         Log.e("GardenViewModel", "Owned plants couldn't be retrieved from repository", e)
         setErrorMsg("getAllPlants failed : Owned plants couldn't be retrieved from repository")
       }
     }
   }
+
+  /**
+   * Fetches all needed user information from the [profileRepo] or set an error message if the fetch
+   * failed.
+   */
+  private fun fetchProfileInfos() {
+    viewModelScope.launch {
+      try {
+        profileRepo.getProfile().collect { profile ->
+          if (profile != null) {
+            _uiState.value =
+                _uiState.value.copy(userName = profile.firstName, userAvatar = profile.avatar)
+          } else {
+            setErrorMsg("Failed to get user profile")
+          }
+        }
+      } catch (_: Exception) {}
```

You could display a Toast if you catch an exception at least, otherwise it would be hard to debug because here the exception is only swallowed by the catch without explanations.

---

Done.